metadata/src/lib.rs

```diff
@@ -0,0 +1,529 @@
+    pallets.sort_by_key(|key| key.1);
+
+    // Note: pallet name is excluded from hashing.
+    // Each pallet has a hash of 32 bytes, and the vector is extended with
+    // extrinsic hash and metadata ty hash (2 * 32).
+    let mut bytes = Vec::with_capacity(pallets.len() * 32 + 64);
+    for (_, hash) in pallets.iter() {
+        bytes.extend(hash)
+    }
+
+    bytes.extend(get_extrinsic_hash(&metadata.types, &metadata.extrinsic));
+
+    let mut visited_ids = HashSet::<u32>::new();
+    bytes.extend(get_type_hash(
+        &metadata.types,
+        metadata.ty.id(),
+        &mut visited_ids,
+    ));
+
+    let hash = hash(&bytes);
+
+    MetadataHashDetails {
+        metadata_hash: hash,
+        pallet_hashes: pallets.into_iter().collect(),
+    }
+}
+
+/// Metadata hasher internal cache.
+#[derive(Clone, Debug)]
+pub struct MetadataHashDetails {
+    /// Full metadata hash.
+    pub metadata_hash: [u8; 32],
+    /// Pallet hashes resulted from metadata hashing.
+    pub pallet_hashes: HashMap<String, [u8; 32]>,
+}
+
+impl MetadataHashDetails {
+    /// Creates an empty `MetadataHasherCache`.
+    pub fn new() -> MetadataHashDetails {
+        Self {
+            metadata_hash: [0; 32],
+            pallet_hashes: HashMap::new(),
+        }
+    }
+}
+
+impl Default for MetadataHashDetails {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        get_metadata_hash,
+        get_pallet_hash,
+    };
+    use frame_metadata::{
+        ExtrinsicMetadata,
+        PalletCallMetadata,
+        PalletMetadata,
+        RuntimeMetadataLastVersion,
+    };
+    use scale_info::meta_type;
+
+    // Define recursive types.
+    #[allow(dead_code)]
+    #[derive(scale_info::TypeInfo)]
+    struct A {
+        pub b: Box<B>,
+    }
+    #[allow(dead_code)]
+    #[derive(scale_info::TypeInfo)]
+    struct B {
+        pub a: Box<A>,
+    }
+
+    fn build_default_extrinsic() -> ExtrinsicMetadata {
+        ExtrinsicMetadata {
+            ty: meta_type::<()>(),
+            version: 0,
+            signed_extensions: vec![],
+        }
+    }
+
+    fn build_default_pallet() -> PalletMetadata {
+        PalletMetadata {
+            name: "Test",
+            storage: None,
+            calls: None,
+            event: None,
+            constants: vec![],
+            error: None,
+            index: 0,
+        }
+    }
+
+    fn build_default_pallets() -> Vec<PalletMetadata> {
+        let mut pallet_first = build_default_pallet();
+        pallet_first.name = "First";
+        pallet_first.calls = Some(PalletCallMetadata {
+            ty: meta_type::<u8>(),
+        });
+        let mut pallet_second = build_default_pallet();
+        pallet_second.name = "Second";
+        pallet_second.calls = Some(PalletCallMetadata {
+            ty: meta_type::<u16>(),
+        });
+        pallet_second.index = 1;
+
+        vec![pallet_first, pallet_second]
```

Comment:
Nit: this might be clearer if you just inline the `PalletMetadata` stuff; something like:

```rust
fn build_default_pallets() -> Vec<PalletMetadata> {
  vec![
    PalletMetadata {
       name: "First",
       // ...
    },
    PalletMetadata {
       name: "Second",
       // ...
    }
  ]
}
```

I don't feel like the function to create a default instance that you then need to modify really pulls it's weight.

Orr, you keep it and construct something like

```
let pallet_first = PalletMetadata {
    name: "First",
    index: 1,
    calls: //... 
    ..default_pallet()
}
```

Reply:
Have picked the latter version with a comprehensive test type :D Thanks